### PR TITLE
Fix SSC Battery State Issue

### PR DESCRIPTION
**Pull request by Myoldmopar**
Pull request overview
---------------------
 - Fixes an issue with the SSC model where last_idx was running wild.  

FYI @bonnema @dguittet

### CODE REVIEW

**Comment by Myoldmopar** [comment]
@dguittet I know that @bonnema was still seeing some indeterminate behavior; I'm wondering if there is still another uninitialized variable like the one I fixed.  I'll be running this under valgrind to find out.

File: src/EnergyPlus/ElectricPowerServiceManager.hh [hunk 1]
```diff
@@ -440,7 +440,9 @@ private: // data
     // Li-ion NMC battery objects from SAM Simulation Core lib_battery
     std::unique_ptr<battery_t> ssc_battery_;
     std::unique_ptr<battery_state> ssc_lastBatteryState_;
+    Real64 ssc_lastBatteryTimeStep_;
```
Comment (dguittet) [inline src/EnergyPlus/ElectricPowerServiceManager.hh]: @Myoldmopar Perhaps this should be initialized too

**Comment by Myoldmopar** [comment]
Odd.  Valgrind isn't actually revealing memory issues here.  There are some uninitialized memory warnings deep deep in the re2 library, but once I suppress those, I don't see any happening inside E+ itself with the defect file.  I'm going to just update that variable above and see if I notice any others, and commit.  @bonnema you are welcome to make another test package if you want to play with it.

**Comment by bonnema** [comment]
I am still getting the same error. @EnergyArchmage, any ideas?
```
INVERTER_PRGM,Line 13,SET INV_REACT_PWR = S * @SIN(THETA),0.0, During Warmup, Occurrence info=RUNPERIOD 1, 01/01 00:00 - 00:10
INVERTER_PRGM,Line 14,SET BATT_CURVE_ACT = INV_ACT_PWR, *** Error: EvaluateExpression: Variable = 'INV_ACT_PWR' used in expression has not been initialized! *** , During Warmup, Occurrence info=RUNPERIOD 1, 01/01 00:00 - 00:10
```

**Comment by Myoldmopar** [comment]
@mjwitte @EnergyArchmage OK, so that 2 line snippet does look bad, but here is the full program:

```
EnergyManagementSystem:Program,
  Inverter_Prgm,                          !- Name
  SET w = battery_discharge_power,        !- Program Line 1
  SET i = battery_current,                !- Program Line 2
  SET v = battery_voltage,                !- Program Line 3
  SET s = i * v,                          !- Program Line 4
  SET pf = 1,                             !- Program Line 5
  SET theta = @ArcCos(pf),                !- Program Line 6
  SET v_grid = 1,                         !- Program Line 7
  SET q = -2.5 * (v_grid - 1),            !- Program Line 8
  IF s <> 0,                              !- Program Line 9
  SET theta = @ArcSin(q/s),               !- Program Line 10
  ENDIF,                                  !- Program Line 11
  SET inv_act_pwr = s * @Cos(theta),      !- Program Line 12
  SET inv_react_pwr = s * @Sin(theta),    !- Program Line 13
  SET batt_curve_act = inv_act_pwr;       !- Program Line 14
```

Note that he is, in fact, setting inv_act_pwr on line 12; and then trying to use it on 14.  And it acts as if that variable never existed.

image


The full EDD shows this expanded out for the whole program:

```
INVERTER_PRGM,Line 1,SET W = BATTERY_DISCHARGE_POWER,0.0, During Warmup, Occurrence info=RUNPERIOD 1, 01/01 00:00 - 00:10
INVERTER_PRGM,Line 2,SET I = BATTERY_CURRENT,0.0, During Warmup, Occurrence info=RUNPERIOD 1, 01/01 00:00 - 00:10
INVERTER_PRGM,Line 3,SET V = BATTERY_VOLTAGE,0.0, During Warmup, Occurrence info=RUNPERIOD 1, 01/01 00:00 - 00:10
INVERTER_PRGM,Line 4,SET S = I * V,0.0, During Warmup, Occurrence info=RUNPERIOD 1, 01/01 00:00 - 00:10
INVERTER_PRGM,Line 5,SET PF = 1,1.000000, During Warmup, Occurrence info=RUNPERIOD 1, 01/01 00:00 - 00:10
INVERTER_PRGM,Line 6,SET THETA = @ARCCOS(PF),0.0, During Warmup, Occurrence info=RUNPERIOD 1, 01/01 00:00 - 00:10
INVERTER_PRGM,Line 7,SET V_GRID = 1,1.000000, During Warmup, Occurrence info=RUNPERIOD 1, 01/01 00:00 - 00:10
INVERTER_PRGM,Line 8,SET Q = -2.5 * (V_GRID - 1),0.0, During Warmup, Occurrence info=RUNPERIOD 1, 01/01 00:00 - 00:10
INVERTER_PRGM,Line 9,IF S <> 0,0.0, During Warmup, Occurrence info=RUNPERIOD 1, 01/01 00:00 - 00:10
INVERTER_PRGM,Line 11,ENDIF,, During Warmup, Occurrence info=RUNPERIOD 1, 01/01 00:00 - 00:10
INVERTER_PRGM,Line 12,SET INV_ACT_PWR = S * @COS(THETA),0.0, During Warmup, Occurrence info=RUNPERIOD 1, 01/01 00:00 - 00:10
INVERTER_PRGM,Line 13,SET INV_REACT_PWR = S * @SIN(THETA),0.0, During Warmup, Occurrence info=RUNPERIOD 1, 01/01 00:00 - 00:10
INVERTER_PRGM,Line 14,SET BATT_CURVE_ACT = INV_ACT_PWR, *** Error: EvaluateExpression: Variable = 'INV_ACT_PWR' used in expression has not been initialized! *** , During Warmup, Occurrence info=RUNPERIOD 1, 01/01 00:00 - 00:10
```

What in the world could cause this??  I did do some debugging, but I haven't landed on any issue.  I'll keep trying, but wondered if you had seen this or had any ideas.  Thanks!

**Comment by EnergyArchmage** [comment]
I suspect line 8. Can't just right "-2.5".  No unary minus operator.  0 - 2.5 is how it should be represented.  I would not think zero is the correct evaluation for that in the edd.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by bonnema** [comment]
> I suspect line 8. Can't just right "-2.5". No unary minus operator. 0 - 2.5 is how it should be represented. I would not think zero is the correct evaluation for that in the edd.

Wouldn't line 7 (`SET V_GRID = 1`) cause line 8 (`SET Q = -2.5 * (V_GRID - 1)`) to resolve to zero (`1 - 1 = 0`)?

Regardless, I have removed those lines in the past with the same error. If I change line 13 to `SET INV_REACT_PWR = 1` it runs fine. I have a feeling it has to do with lines 1-3, which carry to line 4. Lines 1-3 are getting sensor values from the battery model, and although the EDD reports zero, I wonder if they're really not really zero. Maybe not being initialized in the C++ code?

**Comment by shorowit** [comment]
@bonnema Have you tried `SET BATT_CURVE_ACT = (INV_ACT_PWR)`? At least, that worked for a [different (?) EMS issue](https://github.com/NREL/EnergyPlus/issues/10279).

**Comment by EnergyArchmage** [comment]
It might not be able to parse the function call @cos() inside the equation.  Should be a separate line to evaluate built-in function to a variable.

Get Outlook for Android<https://aka.ms/AAb9ysg>
________________________________

**Comment by bonnema** [comment]
Thanks, I’ll try these suggestions. Stay tuned.

**Comment by bonnema** [comment]
> It might not be able to parse the function call `@Cos()` inside the equation. Should be a separate line to evaluate built-in function to a variable. 

This appears to be the problem!

**Comment by bonnema** [comment]
Here is the final program.
```
EnergyManagementSystem:Program,
  Inverter_Prgm,                          !- Name
  SET w = battery_discharge_power,        !- Program Line 1
  SET i = battery_current,                !- Program Line 2
  SET v = battery_voltage,                !- Program Line 3
  SET s = i * v,                          !- Program Line 4
  SET pf = 1,                             !- Program Line 5
  SET theta = @ArcCos(pf),                !- Program Line 6
  SET v_grid = 1,                         !- Program Line 7
  SET q = (0 - 2.5) * (v_grid - 1),       !- Program Line 8
  IF s <> 0,                              !- Program Line 9
  SET theta = @ArcSin(q/s),               !- Program Line 10
  ENDIF,                                  !- Program Line 11
  SET cos_theta = @Cos(theta),            !- Program Line 12
  SET sin_theta = @Sin(theta),            !- Program Line 13
  SET inv_act_pwr = s * cos_theta,        !- Program Line 14
  SET inv_react_pwr = s * sin_theta,      !- Program Line 15
  SET batt_curve_act = (inv_act_pwr);     !- Program Line 16
```
I performed the modifications in steps, first Line 8 (`(0 - 2.5) * (v_grid - 1)`) and then Line 16 (adding parenthesis around the variable). Still had the same error. However, adding Line 12 and 13 (a separate line for the sine and cosine function) resulted in the simulation completing.

The EDD looks correct as well:
```
INVERTER_PRGM,Line 1,SET W = BATTERY_DISCHARGE_POWER,0.0, Occurrence info=RUNPERIOD 1, 12/31 23:50 - 24:00
INVERTER_PRGM,Line 2,SET I = BATTERY_CURRENT,0.0, Occurrence info=RUNPERIOD 1, 12/31 23:50 - 24:00
INVERTER_PRGM,Line 3,SET V = BATTERY_VOLTAGE,532.359919, Occurrence info=RUNPERIOD 1, 12/31 23:50 - 24:00
INVERTER_PRGM,Line 4,SET S = I * V,0.0, Occurrence info=RUNPERIOD 1, 12/31 23:50 - 24:00
INVERTER_PRGM,Line 5,SET PF = 1,1.000000, Occurrence info=RUNPERIOD 1, 12/31 23:50 - 24:00
INVERTER_PRGM,Line 6,SET THETA = @ARCCOS(PF),0.0, Occurrence info=RUNPERIOD 1, 12/31 23:50 - 24:00
INVERTER_PRGM,Line 7,SET V_GRID = 1,1.000000, Occurrence info=RUNPERIOD 1, 12/31 23:50 - 24:00
INVERTER_PRGM,Line 8,SET Q = (0 - 2.5) * (V_GRID - 1),0.0, Occurrence info=RUNPERIOD 1, 12/31 23:50 - 24:00
INVERTER_PRGM,Line 9,IF S <> 0,0.0, Occurrence info=RUNPERIOD 1, 12/31 23:50 - 24:00
INVERTER_PRGM,Line 11,ENDIF,, Occurrence info=RUNPERIOD 1, 12/31 23:50 - 24:00
INVERTER_PRGM,Line 12,SET COS_THETA = @COS(THETA),1.000000, Occurrence info=RUNPERIOD 1, 12/31 23:50 - 24:00
INVERTER_PRGM,Line 13,SET SIN_THETA = @SIN(THETA),0.0, Occurrence info=RUNPERIOD 1, 12/31 23:50 - 24:00
INVERTER_PRGM,Line 14,SET INV_ACT_PWR = S * COS_THETA,0.0, Occurrence info=RUNPERIOD 1, 12/31 23:50 - 24:00
INVERTER_PRGM,Line 15,SET INV_REACT_PWR = S * SIN_THETA,0.0, Occurrence info=RUNPERIOD 1, 12/31 23:50 - 24:00
INVERTER_PRGM,Line 16,SET BATT_CURVE_ACT = (INV_ACT_PWR),0.0, Occurrence info=RUNPERIOD 1, 12/31 23:50 - 24:00
```
Thanks for all your help @Myoldmopar, @EnergyArchmage, @mjwitte, @shorowit, and @dguittet!

**Comment by EnergyArchmage** [comment]
note that the syntax for functions is space delimitated arguments.  I see lots of parentheses that would not expect to work but apparently they are getting parsed away.  Would have to run in debugger to see how the tokens are really getting parsed. 

Line 10 seems problematic, the q/s should be evaluated separately and then passed as an argument.

**Comment by bonnema** [comment]
> note that the syntax for functions is space delimitated arguments. I see lots of parentheses that would not expect to work but apparently they are getting parsed away. Would have to run in debugger to see how the tokens are really getting parsed.
> 
> Line 10 seems problematic, the q/s should be evaluated separately and then passed as an argument.

I see. I think during my testing `s` is always zero, and the line 9 `IF` is never true. I'll split the `q/s`, and remove some of the parenthesis. Thanks again!

**Comment by Myoldmopar** [comment]
This is great news.  I pulled in develop to get clean CI results and we can still plan on merging this initialization fix.

**Comment by Myoldmopar** [comment]
That's more like it, just a diff in one SSC file.  That's what I see locally as well.  I'll let CI finish, but otherwise this is good to go.

**Comment by Myoldmopar** [comment]
All clean, merging.

**Pull request merged**